handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandlerBuilder.java

```diff
@@ -0,0 +1,107 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.proxy;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+
+import io.netty.util.internal.StringUtil;
+
+/**
+ * Creates a new {@link ProxyHandler} of desired type with requested proxy
+ * address and authentication. The builder gracefully accepts null or empty
+ * values for username and password, which allows callers to pass in values from
+ * configuration files without additional checking.
+ */
+public final class ProxyHandlerBuilder {
+
+  public static enum ProxyType {
+    HTTP, SOCKS4, SOCKS5
+  }
+
+  public static ProxyHandlerBuilder forType(ProxyType proxyType) {
+    return new ProxyHandlerBuilder(proxyType);
+  }
+
+  public static ProxyHandlerBuilder forType(String proxyType) {
+    for (ProxyType type : ProxyType.values()) {
+      if (type.name().equalsIgnoreCase(proxyType)) {
+        return new ProxyHandlerBuilder(type);
+      }
+    }
+    throw new RuntimeException("Proxy type " + proxyType + " not supported.");
+  }
+
+  private final ProxyType type;
+  private SocketAddress proxyAddress;
+  private String username;
+  private String password;
+
+  private ProxyHandlerBuilder(ProxyType type) {
+    this.type = type;
+  }
+
+  public ProxyHandlerBuilder proxyAddress(SocketAddress proxyAddress) {
+    checkNotNull(proxyAddress, "proxyAddress");
+    this.proxyAddress = proxyAddress;
+    return this;
+  }
+
+  public ProxyHandlerBuilder proxyAddress(String host, int port) {
+    checkNotNull(host, "proxyHost");
+    this.proxyAddress = InetSocketAddress.createUnresolved(host, port);
+    return this;
+  }
+
+  public ProxyHandlerBuilder username(String username) {
+    this.username = username;
+    return this;
+  }
+
+  public ProxyHandlerBuilder password(String password) {
+    this.password = password;
+    return this;
+  }
+
+  public ProxyHandler build() {
+    checkNotNull(proxyAddress, "proxyAddress");
+    switch (type) {
+    case HTTP:
+      if (StringUtil.isNullOrEmpty(username)) {
+        return new HttpProxyHandler(proxyAddress);
+      } else {
+        return new HttpProxyHandler(proxyAddress, username, password);
+      }
+    case SOCKS4:
+      if (StringUtil.isNullOrEmpty(username)) {
+        return new Socks4ProxyHandler(proxyAddress);
+      } else {
+        return new Socks4ProxyHandler(proxyAddress, username);
+      }
+    case SOCKS5:
+      if (StringUtil.isNullOrEmpty(username)) {
+        return new Socks5ProxyHandler(proxyAddress);
+      } else {
+        return new Socks5ProxyHandler(proxyAddress, username, password);
+      }
+    default:
+      throw new RuntimeException("Proxy type " + type + " not supported.");
```

Comment:
you really use java.lang.Error for that? While it is a problem which is similar to noclassdeferror, i find that pretty hard. done anyway :)